### PR TITLE
Update jira-time-tracking extension

### PR DESCRIPTION
**Pull request by haydencbarnes**
## Description

1. Originally, the extension only supported Jira Cloud users. This PR adds support for Jira Server users as well. Fix for: https://github.com/raycast/extensions/issues/11914
2. I have updated the time logging UI/UX to be more Raycast-like; see the gif below. (including adding regex only to accept acceptable data on the frontend to be parsed.)

## Screencast
1.
CleanShot 2024-04-21 at 12 58 58@2x
Also, I added a new selector for users to select while setting up the extension if they are using Jira Server and not Cloud. Note: Assuming most users will use Jira Cloud as default as Jira Server support from Atlassian has been winding down.

2.
CleanShot 2024-04-21 at 11 42 19
Updated Time Logging UI/UX

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

**Comment by haydencbarnes** [comment]
@niallpaterson, I would appreciate it if you could test this! After posting this, I realized that I do not have cloud permissions on my side besides looking at projects. (So I have verified it receives Jira API v3 data, but I cannot add time to any tickets, so that was not tested yet, unfortunately.)

**Comment by haydencbarnes** [comment]
@0w0miki Would also greatly appreciate a review from you as I changed some of the newer pagination code you had written!

**Comment by niallpaterson** [comment]
Hey @haydencbarnes,

Thanks for your contribution! 

Unfortuantely I do not have access to Jira anymore so cannot test, I'm afraid. But the code looks good to me.

One small UI point: it feels odd to opt out with the checkbox. I would have thought you would opt into using Jira Server (i.e., the checkbox is unchecked by default.) Anyways, hair splitting really.

**Comment by haydencbarnes** [comment]
> One small UI point: it feels odd to opt out with the checkbox. I would have thought you would opt into using Jira Server (i.e., the checkbox is unchecked by default.) Anyways, hair splitting really.

Yeah, I was trying to decide on the best way to do that, I might turn it into a dropdown here in a minute or so.

**Comment by haydencbarnes** [comment]
Added a selector @niallpaterson, believe that is better UI

**Comment by pernielsentikaer** [comment]
@niallpaterson do you mind checking one last time, I don't have a cloud account either but users will report bugs if there is something breaking

**Comment by haydencbarnes** [comment]
Found a bug with entering time where the time was not reset to 0 after first time entered, recent commit fixes that

**Comment by haydencbarnes** [comment]
@0w0miki I am pinging you again to see if you can test that the cloud is still working as expected with this change?

**Comment by pernielsentikaer** [comment]
Let's give @0w0miki a few days to respond

**Comment by 0w0miki** [comment]
Hi @haydencbarnes,
I just got time to review this today. Actually, I am not familiar with React. Therefore, I need some time for this.

**Review by 0w0miki — COMMENTED** [review]
I have made a test with Jira Cloud. Only the first 50 projects and 100 issues can be chosen.

File: extensions/jira-time-tracking/src/index.tsx [hunk 1]
```diff
@@ -1,105 +1,157 @@
...
+      parseTimeInput(newTime);
+    }
+  };
 
-    if (!selectedIssue) {
-      showToast(Toast.Style.Failure, "Error logging time: issue not found");
+  const parseTimeInput = (input: string) => {
+    const totalSeconds = parseTimeToSeconds(input);
+
+    if (totalSeconds < 900) {
```
[the remark this answers is not in the document]
Comment (haydencbarnes) [reply extensions/jira-time-tracking/src/index.tsx:39]: > I don't think it is a good idea to add a limitation that is not existing on Jira web/API. If there is requirement for minimum time to log, I prefer to provide a setting for user to configure.

Good points, will add the minimum time to log as a UI setting in the future and leave it alone for now.

[the remark this answers is not in the document]
Comment (haydencbarnes) [reply extensions/jira-time-tracking/src/index.tsx:39]: > Besides, logically, the bound validation seems to be a part of `isValidTimeInput()`.

Yeah, I should have cleaned this up/moved it to the correct place if it was a real requirement, good point.

[the remark this answers is not in the document]
Comment (haydencbarnes) [reply extensions/jira-time-tracking/src/index.tsx]: [98e95d2](https://github.com/raycast/extensions/pull/11918/commits/98e95d22fa48189ec8e6908e5dc00c8f6ac332b0) Should fix this.

File: extensions/jira-time-tracking/src/index.tsx [hunk 2]
```diff
@@ -1,105 +1,157 @@
...
+      } finally {
+        setLoading(false);
       }
+    };
+
+    if (projects.length === 0 || pageGot.current < pageTotal.current) {
+      fetchProjects();
     }
-  }, [projects]);
+  }, [isJiraCloud]);
```
[the remark this answers is not in the document]
Comment (haydencbarnes) [reply extensions/jira-time-tracking/src/index.tsx]: https://github.com/raycast/extensions/pull/11918/commits/cea6a586f9baa9e2a4662fa531300cf4452b4b5c Should fix this

**Comment by haydencbarnes** [comment]
@0w0miki Thank you kindly for the review and good catches! Hopefully the fixes I added should fix the behavior on the cloud side.

**Comment by pernielsentikaer** [comment]
I'm not sure that it's even actual at all, but can we use pagination for [something](https://developers.raycast.com/api-reference/user-interface/grid#pagination)?

**Comment by haydencbarnes** [comment]
@0w0miki if you have the time could you plz review the new changes?

**Comment by 0w0miki** [comment]
I saw you use both `setTimeout()` and states like `issues.length`, `pageGot.current` to trigger project and issue updating. 
This will cause duplicate update on the same item, and the following warning prints.
> Encountered two children with the same key, `TEST`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.

**Comment by 0w0miki** [comment]
> I'm not sure that it's even actual at all, but can we use pagination for [something](https://developers.raycast.com/api-reference/user-interface/grid#pagination)?

It is a good idea to utilize the mechanism provided by Raycast. But the API seems to be only used for Grid component. It looks more like a gallery view of images. I am not sure whether it is a good to use it here, as we do not have any image to display.

**Comment by pernielsentikaer** [comment]
@0w0miki you can also use it for lists, just posted [wrong link](https://developers.raycast.com/api-reference/user-interface/list#pagination) 🙂

**Comment by haydencbarnes** [comment]
> I saw you use both `setTimeout()` and states like `issues.length`, `pageGot.current` to trigger project and issue updating. This will cause duplicate update on the same item, and the following warning prints.
> 
> > Encountered two children with the same key, `TEST`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.

Does the latest commit fix this?

**Comment by pernielsentikaer** [comment]
@0w0miki are you still around?

**Comment by haydencbarnes** [comment]
@pernielsentikaer @0w0miki, I received a Jira Cloud account and tested it on my side. It works as expected on smaller projects. @0w0miki, if you have time to test it on a bigger project with many issues, please let me know if you see any errors.

**Comment by haydencbarnes** [comment]
Last commit here was for any final pagination confusion.

**Comment by pernielsentikaer** [comment]
@0w0miki could you check it a last time?

**Comment by haydencbarnes** [comment]
Latest was fix for https://github.com/raycast/extensions/issues/12550

**Comment by haydencbarnes** [comment]
@0w0miki Are you available to review this?

**Review by 0w0miki — APPROVED** [review]

**Review by pernielsentikaer — APPROVED** [review]
Hi 👋 

Looks good to me, approved 🔥

**Pull request merged**